### PR TITLE
Refactor embed block to single block with block variations

### LINKED ISSUE #22660: Block Library: Refactor Embed block to use `block.json` and block variations
## Description

Part of #16209.

Follow-up for #22491 where all blocks but Embed were refactored to be registered on the server with `block.json` metadata files.

Embed block is really a collection of multiple variations of the same block so ideally it should be converted to one block containing a set of block variations. Similar to what @mcsf did for Social Icon block: #19887.

All that makes it hard to move its settings to a single `block.json` and register in on the server to be exposed in the REST API endpoint with block definitions.

### PR DESCRIPTION
**Pull request by ntsekouras**
Resolves: https://github.com/WordPress/gutenberg/issues/22660

Above issue is part of https://github.com/WordPress/gutenberg/issues/16209 that is a direction toward server-side awareness of block types.

This PR refactors `embed` block to be a single block ( single `block.json` ) with `block variations`, so as to register in on the server to be exposed in the REST API endpoint with block definitions.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-library/src/embed/edit.js [hunk 1]
```diff
@@ -33,181 +35,215 @@ function getResponsiveHelp( checked ) {
...
+			};
+		},
+		[ attributes.url ]
+	);
+
+	/**
+	 * @return {Object} Attributes derived from the preview, merged with the current attributes.
+	 */
+	// TODO maybe use `useCallback` ?? or even better useMemo // or state? with mergedAttributes?
+	const getMergedAttributes = () => {
```
Comment (ntsekouras) [inline packages/block-library/src/embed/edit.js:108]: // should we use `useCallback` or `useMemo` with dependencies here? Would it better for performance?

File: packages/block-library/src/embed/util.js [hunk 2]
```diff
@@ -214,7 +233,7 @@ export const getAttributesFromPreview = memoize(
 		title,
 		currentClassNames,
 		isResponsive,
-		allowResponsive = true
```
Comment (ntsekouras) [inline packages/block-library/src/embed/util.js:217]: // is the default value needed here since it's set in block's attributes?

File: packages/block-library/src/embed/edit.js [hunk 3]
```diff
@@ -33,181 +35,215 @@ function getResponsiveHelp( checked ) {
...
+	}
+
+	// Even though we set attributes that get derived from the preview,
+	// we don't access them directly because for the initial render,
+	// the `setAttributes` call will not have taken effect. If we're
+	// rendering responsive content, setting the responsive classes
+	// after the preview has been rendered can result in unwanted
+	// clipping or scrollbars. The `getAttributesFromPreview` function
+	// that `getMergedAttributes` uses is memoized so that we're not
+	// calculating them on every render.
```
Comment (ntsekouras) [inline packages/block-library/src/embed/edit.js:211]: // Do we need this? Doesn't `setAttributes` cause a rerender, so the new classes would be applied?

File: packages/block-library/src/embed/edit.js (hunk 1 above)
Comment (mcsf) [reply packages/block-library/src/embed/edit.js:108]: Considering that, at the source, `getMergedAttributes` is using `memoize`, then it might be nice to:

* Use `useMemo` here as `const mergedAttributes = useMemo( `
* Remove `memoize` from the definition of `getMergedAttributes`.

**Review by mcsf — COMMENTED** [review]
Nice work so far! I need to come back for another pass.

File: packages/block-library/src/embed/edit.js (hunk 1 above)
Comment (mcsf) [reply packages/block-library/src/embed/edit.js:108]: Actually, I take back what I said. There are a few moving parts here and I'd rather approach this carefully. So I would not touch the `memoize` part of `getAttributesFromPreview`.

As for `useMemo` or other approaches... thinking in abstract, it could be nice to replace the way that we use `getMergedAttributes` with a new piece of state (`useState`). The benefit is that it could remove some indirection. The drawback is that it would introduce different life-cycle considerations (any change in that state would cause a re-render, whereas the effect-based dance is more flexible) which could lead to unforeseen bugs.

**Comment by mcsf** [comment]
One thing we can't forget is to make sure the old types are still supported, i.e. `core/embed-foo`.

One way is to add support at the parser level, as was done for Social Icons:

https://github.com/WordPress/gutenberg/blob/201ca8299e258f535b2a470134dd0ad3cd13de1f/packages/blocks/src/api/parser.js#L432-L439

There's also this comment from Greg about a different way of managing this, but we don't need to solve that now:

https://github.com/WordPress/gutenberg/pull/19887/files#r375181162

**Comment by mcsf** [comment]
I pushed 6d3f5fe to scratch a personal itch: some of these effects become a lot clearer (and more autonomous?) by keeping their specific logic in one place.

File: packages/block-library/src/embed/util.js [hunk 4]
```diff
@@ -227,7 +246,7 @@ export const getAttributesFromPreview = memoize(
 		// because not all embed code gives us a provider name.
 		const { html, provider_name: providerName } = preview;
 		const providerNameSlug = kebabCase(
-			toLower( '' !== providerName ? providerName : title )
+			( providerName ?? title ).toLowerCase()
```
[the remark this answers is not in the document]
Comment (ntsekouras) [reply packages/block-library/src/embed/util.js:259]: You're right. I'll change it to use `||`. It shouldn't pick an `undefined|null` `providerName`.

File: packages/block-library/src/embed/block.json [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+		"providerNameSlug": {
+			"type": "string"
+		},
+		"allowResponsive": {
+			"type": "boolean",
+			"default": true
+		},
+		"responsive": {
+			"type": "boolean",
+			"default": true
```
Comment (mcsf) [inline packages/block-library/src/embed/block.json:25]: Per b0755cc2a3d0cbb3ef27b6c45a80b56ab068e345, don't forget:

```suggestion
			"default": false
```

File: packages/block-library/src/embed/variations.js [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+		transform: [
+			{
+				type: 'block',
+				blocks: [ 'polldaddy' ],
+				transform: ( content ) =>
+					createBlock( 'crowdsignal', {
+						content,
+					} ),
+			},
+		],
```
Comment (mcsf) [inline packages/block-library/src/embed/variations.js:170]: To my knowledge, there are no plans for the Variations API to automatically handle per-variation transforms. And, for now, I think that's a good thing.

For the sake of avoiding complexity, maybe we could take these transforms (which are basically just renamings: `polldaddy -> crowdsignal`, `speaker -> speaker-deck`) and handle this directly in the same place (in the parser?) where we will map `core/embed-foo -> core/embed`.

Comment (ntsekouras) [reply packages/block-library/src/embed/variations.js:170]: I haven't looked into more depth with this issue ( variation transforms ), but your idea definitely sounds goods.

File: packages/e2e-tests/specs/editor/various/embedding.test.js [hunk 7]
```diff
@@ -278,6 +278,6 @@ describe( 'Embedding content', () => {
 		await insertEmbed( postUrl );
 
 		// Check the block has become a WordPress block.
-		await page.waitForSelector( '.wp-block-embed-wordpress' );
+		await page.waitForSelector( 'figure.wp-block-embed' );
```
Comment (ntsekouras) [inline packages/e2e-tests/specs/editor/various/embedding.test.js:281]: // Is there a way to better test some of these tests? Maybe with `is-provider-...` or some other class? I tried //`saveDraft` and `publishPost` and get the element again, but these classes are not there.
// the `is-provider-...` class is added in `save.js`

Comment (mcsf) [reply packages/e2e-tests/specs/editor/various/embedding.test.js:281]: Maybe combine these `waitForSelector` calls with a general `expect` at the end of the test that matches `getEditedPostContent` against a snapshot? As a benefit, this also targets some other pieces that were left unchecked like:

```js
// Photo content. Should render valid figure element.
await insertEmbed( 'https://cloudup.com/cQFlxqtY4ob' );
```

**Comment by mcsf** [comment]
I noticed something when testing this with the Gutenberg demo (`/wp-admin/post-new.php?gutenberg-demo`):

1. Open the demo post
2. Scroll down to the Vimeo embed block
3. Switch to Code mode, then back to Visual

Observed in `master`: no change in the embed's alignment and size:

before

Observed in this branch: the embed looks fine before switching modes, but not after:

after-2

**Comment by cameronvoell** [comment]
On the native mobile side, we do not support any embed blocks yet, but we did see an issue after this change related to how we direct users to edit unsupported (native mobile) blocks from the mobile editor. A GIF demonstrating that flow can be seen on [this PR](https://github.com/WordPress/gutenberg/pull/23592). We left a note with mobile devs who worked on that flow and should have more info soon on necessary mobile side changes.

**Comment by hypest** [comment]
> We left a note with mobile devs who worked on that flow and should have more info soon on necessary mobile side changes.

Thanks for the ping @ntsekouras and for looking into this @cameronvoell ! If possible Nik, can we momentarily hold merging this PR until we get a better idea on how to fix/avoid the mobile breakage (@etoledom is looking deeper into this)? Let me know if there is any kind of urgency for landing this PR though, thanks!

**Comment by ntsekouras** [comment]
> I noticed something when testing this with the Gutenberg demo (/wp-admin/post-new.php?gutenberg-demo):

Good catch @mcsf! It was missing the `responsive` property that is now is set to `false` by default. I fixed it.

**Comment by ntsekouras** [comment]
Hey @hypest. Thank you all who took the time to look into this PR. Maybe there is anything I can do to help you out?

-cc @cameronvoell @etoledom

**Comment by mcsf** [comment]
> If possible Nik, can we momentarily hold merging this PR until we get a better idea on how to fix/avoid the mobile breakage

Hey, @hypest! There's no urgency on the feature itself and we can pause. However, this branch touches many things and is a significant change in how embeds work, so our main worry is being stuck having to keep this branch fresh against other changes. I'd prefer it if merging weren't held for more than a few days. What do you think?

**Comment by etoledom** [comment]
> On the native mobile side, we do not support any embed blocks yet, but we did see an issue after this change related to how we direct users to edit unsupported (native mobile) blocks from the mobile editor. A GIF demonstrating that flow can be seen on [this PR](https://github.com/WordPress/gutenberg/pull/23592). We left a note with mobile devs who worked on that flow and should have more info soon on necessary mobile side changes.

After checking this out in more details, we came to the conclusion that the experimented issue was due to a problem in wpcom, that is generating a block markup that is not recognised by the migration implemented on this PR.

So it's safe to say that this PR in particular doesn't bring regressions to the Native implementation of the editor. 🎊

Thank you @ntsekouras and @mcsf for minding the mobile!
cc @cameronvoell @hypest

**Comment by mcsf** [comment]
I noticed this branch is affecting the appender:

Before|After
-|-
Captura de ecrã 2020-07-30, às 12 53 24|Captura de ecrã 2020-07-30, às 12 52 55

File: packages/block-editor/src/components/inserter/utils.js [hunk 8]
```diff
@@ -1,45 +1,71 @@
...
+ * @param {number} [limit] Limit returned results by a given threshold. If limit===-1 return all
  * @return {Array} Normalized inserter items.
  */
-export function includeVariationsInInserterItems( items ) {
-	return items.reduce( ( result, item ) => {
-		const { variations = [] } = item;
-		const hasDefaultVariation = variations.some(
-			( { isDefault } ) => isDefault
-		);
+export function includeVariationsInInserterItems( items, limit = -1 ) {
```
Comment (mcsf) [inline packages/block-editor/src/components/inserter/utils.js:26]: 1. There's some inconsistency between this default of `-1` and the subsequent checks against `0`. Since this isn't an `index`-type function, I suggest `0` instead of `-1`. Another option totally valid in JS would be `limit = Infinity`, which is kinda cool.

2. The inclusion check below reads `limit > 0` but its counterpart reads `limit < 1`. To avoid surprises, I recommend `limit <= 0` (or just `limit === 0`, `! limit`, whichever you prefer, since we should move away from the `-1` default).

File: packages/block-editor/src/components/inserter/utils.js [hunk 9]
```diff
@@ -1,45 +1,71 @@
...
-						initialAttributes: {
-							...item.initialAttributes,
-							...variation.attributes,
-						},
-						innerBlocks: variation.innerBlocks,
-					};
-				} )
+			if ( variations.length ) {
+				const variationMapper = getVariationMapper( item );
+				result = result.concat( variations.map( variationMapper ) );
```
Comment (mcsf) [inline packages/block-editor/src/components/inserter/utils.js:48]: Given the circumstances, and the fact that we already do it a few lines above, let's replace concat:

```suggestion
				result.push( ...variations.map( variationMapper ) );
```

It [should be safe](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply):

> The consequences of applying a function with too many arguments (that is, **more than tens of thousands** of arguments) varies across engines. (The JavaScriptCore engine has hard-coded argument limit of 65536.

:D

File: packages/block-editor/src/components/inserter/utils.js [hunk 10]
```diff
@@ -1,45 +1,71 @@
+// TODO jsdoc + tests
+const getVariationMapper = ( item ) => ( variation ) => ( {
```
Comment (mcsf) [inline packages/block-editor/src/components/inserter/utils.js:2]: I used to be a huge fan of currying and other FP techniques in JS. :) In the context of Gutenberg, I'm a bit more conservative. Plus, arrow functions mitigate some of the need.

1. Anyway, the name is probably what needs the most work. `getItemFromVariation`? wdyt?

2. As for `( item ) => ( variation ) =>` vs. `( item, variation ) =>`, it's ultimately your preference.

File: packages/block-editor/src/components/inserter/utils.js [hunk 11]
```diff
@@ -1,45 +1,71 @@
...
+			}
+
+			return result;
+		}, [] );
+	}
+
+	// Limit is bigger than items length so we can add variations if exist
+	let itemsToAdd = limit - itemsLength;
+	let variationsToAdd = [];
+	for ( const item of items ) {
```
Comment (mcsf) [inline packages/block-editor/src/components/inserter/utils.js:58]: In practice, how likely are we to hit this scenario?

The solution chosen here is to basically backfill `items` in no "meaningful" order (just iterating on `items`) until there happen to be enough variations added to `items` that we can break out and serve the result. Which isn't really helpful to users, IMO (but, also, I'm not sure when we will ever have a `limit` so high that we ever need to backfill).

So I recommend the following:

* For the purposes of this PR and fixing the Quick Inserter issue, drop this part entirely (just keep the essential logic around `items.slice( 0, limit )`).

* For a "proper" long-term solution, we should fix this (at the selector level?) so that variations usage is counted and variations can be properly ordered in usage along with the rest of the inserter items.

Does that seem reasonable?

File: packages/block-editor/src/components/inserter/utils.js (hunk 8 above)
Comment (ntsekouras) [reply packages/block-editor/src/components/inserter/utils.js:26]: > this default of -1
>  would be limit = Infinity

I had thought about `Infinity` too but I used `-1` with `$posts_per_page` in my mind. Let's go with `Infinity` then.

Regarding your second comment, the checks will change.

File: packages/block-editor/src/components/inserter/utils.js (hunk 11 above)
Comment (ntsekouras) [reply packages/block-editor/src/components/inserter/utils.js:58]: > In practice, how likely are we to hit this scenario?

I believe `zero` but I had to make a solution to handle it, as I wanted feedback :).

> drop this part entirely (just keep the essential logic around items.slice( 0, limit )).

I agree with you on this. I'll change it.

> For a "proper" long-term solution, we should fix this (at the selector level?) so that variations usage is counted and variations can be properly ordered in usage along with the rest of the inserter items.

Actually the problem lies in the fact that we add every available variation to be used in other parts of inserter, like searching to insert a block. So this happens after the `selector` stage.

A new selector for variations' usage would be one part of the solution and the second step would be to handle the addition of variation better or more gracefully, maybe with some `config` or even extract that logic to be used when needed.

**Comment by GhBsx-png** [comment]
Thank you for your help!

File: post-content.php [hunk 12]
```diff
@@ -145,11 +145,11 @@
 <p><?php _e( 'Any block can opt into these alignments. The embed block has them also, and is responsive out of the box:', 'gutenberg' ); ?></p>
 <!-- /wp:paragraph -->
 
-<!-- wp:core-embed/vimeo {"url":"https://vimeo.com/22439234","type":"video","providerNameSlug":"vimeo","align":"wide","className":"wp-has-aspect-ratio wp-embed-aspect-16-9"} -->
-<figure class="wp-block-embed-vimeo alignwide wp-block-embed is-type-video is-provider-vimeo wp-has-aspect-ratio wp-embed-aspect-16-9"><div class="wp-block-embed__wrapper">
```
Comment (mcsf) [inline post-content.php:149]: Hm, are we removing these classes (`wp-block-embed-${type}`) from the block output? We should keep them, they are important for compat.

File: packages/e2e-tests/specs/editor/various/embedding.test.js (hunk 7 above)
Comment (mcsf) [reply packages/e2e-tests/specs/editor/various/embedding.test.js:284]: Now looking at these files again, I notice the lack of `wp-block-embed-${type}` classes, as I pointed out in the other comment. Sorry I didn't point this out earlier.

File: packages/e2e-tests/fixtures/blocks/core__embed.json [hunk 13]
```diff
@@ -6,7 +6,9 @@
         "attributes": {
             "url": "https://example.com/",
             "caption": "Embedded content from an example URL",
-            "allowResponsive": true
+			"allowResponsive": true,
+			"responsive": false,
+			"previewable": true
```
Comment (mcsf) [inline packages/e2e-tests/fixtures/blocks/core__embed.json:11]: Weird, why is the indentation different?

File: packages/blocks/src/api/serializer.js [hunk 14]
```diff
@@ -36,7 +36,7 @@ import BlockContentProvider from '../block-content-provider';
  */
 export function getBlockDefaultClassName( blockName ) {
 	// Generated HTML classes for blocks follow the `wp-block-{name}` nomenclature.
-	// Blocks provided by WordPress drop the prefixes 'core/' or 'core-' (used in 'core-embed/').
+	// Blocks provided by WordPress drop the prefixes 'core/' or 'core-'.
```
Comment (mcsf) [inline packages/blocks/src/api/serializer.js:39]:
```suggestion
	// Blocks provided by WordPress drop the prefixes 'core/' or 'core-' (historically used in 'core-embed/').
```

File: packages/blocks/src/api/parser.js [hunk 15]
```diff
@@ -438,6 +438,28 @@ export function createBlockWithFallback( blockNode ) {
...
+	if ( name && name.indexOf( 'core-embed/' ) === 0 ) {
+		// Capture `core-embed/instagram` into `{"providerNameSlug":"instagram"}`
+		const providerSlug = name.substring( 11 );
+		const deprecated = {
+			speaker: 'speaker-deck',
+			polldaddy: 'crowdsignal',
+		};
+		attributes.providerNameSlug = Object.keys( deprecated ).includes(
+			providerSlug
+		)
```
Comment (mcsf) [inline packages/blocks/src/api/parser.js:452]: Instead of `Object.keys( a ).includes( b )`, might as well use `b in a` or `a.hasOwnProperty( b )`.

File: packages/block-library/src/embed/save.js [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+		return null;
+	}
+
+	const embedClassName = classnames( 'wp-block-embed', {
+		[ `is-type-${ type }` ]: type,
+		[ `is-provider-${ providerNameSlug }` ]: providerNameSlug,
+	} );
+
+	return (
+		<figure className={ embedClassName }>
```
Comment (mcsf) [inline packages/block-library/src/embed/save.js:24]: In `master`, if I paste a YouTube link and look at the markup of the generated embed, I see:

```html
<figure class="wp-block-embed-youtube wp-block-embed is-type-video is-provider-youtube wp-embed-aspect-16-9 wp-has-aspect-ratio">
```

The following classes seem to be missing in this branch:

* `wp-block-embed-${ providerNameSlug }`
* `wp-embed-aspect-16-9`
* `wp-has-aspect-ratio`

**Review by mcsf — CHANGES_REQUESTED** [review]
Almost there! But the consistency/compatibility of the block markup, esp. HTML classes, is really important to get right.

File: post-content.php (hunk 12 above)
Comment (ntsekouras) [reply post-content.php:149]: I'll change it back - is this needed for possible theme style overrides?

File: packages/e2e-tests/fixtures/blocks/core__embed.json (hunk 13 above)
Comment (ntsekouras) [reply packages/e2e-tests/fixtures/blocks/core__embed.json:11]: It had spaces so I changed to `tabs`

File: packages/block-library/src/embed/save.js (hunk 16 above)
Comment (ntsekouras) [reply packages/block-library/src/embed/save.js:25]: > wp-embed-aspect-16-9
wp-has-aspect-ratio

These classes seem to work just fine. Can you share your testing link?

> wp-block-embed-${ providerNameSlug }

Previously this class was added before save as default block name class. With your previous comment for backwards compatibility I added it in `save` function. So it will not be visible before save but the final classes are identical (after save).

I believe this is okay, no?

File: post-content.php (hunk 12 above)
Comment (mcsf) [reply post-content.php:149]: Yeah, that's the idea.

File: packages/block-library/src/embed/save.js (hunk 16 above)
Comment (mcsf) [reply packages/block-library/src/embed/save.js:25]: > These classes seem to work just fine. Can you share your testing link?

I don't see it now, I'm not sure how I tested. :) 👍 

> I believe this is okay, no?

Yep!

File: packages/block-library/src/embed/util.js [hunk 17]
```diff
@@ -174,8 +173,8 @@ export function getClassNames(
 			const potentialRatio = ASPECT_RATIOS[ ratioIndex ];
 			if ( aspectRatio >= potentialRatio.ratio ) {
 				return classnames( existingClassNames, {
-					[ potentialRatio.className ]: allowResponsive,
-					'wp-has-aspect-ratio': allowResponsive,
+					[ potentialRatio.className ]: true,
+					'wp-has-aspect-ratio': true,
```
Comment (mcsf) [inline packages/block-library/src/embed/util.js:177]: What should we know about this change?

File: packages/e2e-tests/specs/editor/various/embedding.test.js [hunk 18]
```diff
@@ -171,7 +171,7 @@ describe( 'Embedding content', () => {
 	it( 'should render embeds in the correct state', async () => {
 		// Valid embed. Should render valid figure element.
 		await insertEmbed( 'https://twitter.com/notnownikki' );
-		await page.waitForSelector( 'figure.wp-block-embed-twitter' );
+		await page.waitForSelector( 'figure.wp-block-embed' );
```
Comment (mcsf) [inline packages/e2e-tests/specs/editor/various/embedding.test.js:174]: Now that we are sure to preserve `wp-block-embed-xyz` classes, maybe we can revert the changes in this test file and remove the snapshots? In general, it gives me confidence that we're able to preserve these tests throughout a refactor :)

File: packages/block-library/src/embed/util.js (hunk 17 above)
Comment (ntsekouras) [reply packages/block-library/src/embed/util.js:177]: At the start of this function there is a check:
```
if ( ! allowResponsive ) { .... return ...; }
```

So this is the path for being responsive and doesn't need to take the value from variable. And I believe this is clearer to what this code does.

Comment (mcsf) [reply packages/block-library/src/embed/util.js:177]: You're totally right, I see it now.

Btw, a tip: `classNames` accepts as many strings as needed, so in this case where we always have `true` we could rewrite as:

```js
classNames( existingClassNames, potentialRatio.className, 'wp-has-aspect-ratio' )
```

File: packages/e2e-tests/specs/editor/various/embedding.test.js (hunk 18 above)
Comment (ntsekouras) [reply packages/e2e-tests/specs/editor/various/embedding.test.js:174]: I believe that having the snapshot isn't a bad thing here. In addition:
>Previously this class was added before save as default block name class. With your previous comment for backwards compatibility I added it in save function. So it will not be visible before save but the final classes are identical (after save).

If you think this should really change and remove the snapshots I will have to change the tests to save the post and then check for the class. Just tell me to change it this will be the direction.

Comment (mcsf) [reply packages/e2e-tests/specs/editor/various/embedding.test.js:174]: OK, not worth the effort then :)

File: packages/block-library/src/embed/util.js (hunk 17 above)
Comment (ntsekouras) [reply packages/block-library/src/embed/util.js]: Good tip! Thanks! Changed it.

**Review by mcsf — APPROVED** [review]
Great job here!

**Comment by ockham** [comment]
It _seems_ to me that `preview` is always `undefined` in `createUpgradedEmbedBlock`. This is only relevant for WordPress embeds (since they're the only embed that doesn't just look at the URL but the preview HTML). IIRC, the `preview` used to be generated through some component lifecycle methods; maybe something there got broken.

(Please take with a grainof salt, this is the result of some preliminary testing; I'll continue to look into this tomorrow.)

**Comment by ockham** [comment]
ping @ntsekouras about my https://github.com/WordPress/gutenberg/pull/24090#issuecomment-671564833:

> It _seems_ to me that `preview` is always `undefined` in `createUpgradedEmbedBlock`. This is only relevant for WordPress embeds (since they're the only embed that doesn't just look at the URL but the preview HTML). IIRC, the `preview` used to be generated through some component lifecycle methods; maybe something there got broken.

Can you have a look what might've changed there? :pray: 

---

Furthermore, related to this comment: https://github.com/Automattic/wp-calypso/issues/39935#issuecomment-671825119:

> There is proper backwards compatibility with `core` but I've noticed in `.com` the above field `providerNameSlug` is not filled by some middleware you have there, I guess. This will cause problems with older `embed` blocks as it compares them with current version and the `providerNameSlug` adds `css classes` based on provider, that makes them invalid.
> 
> So you'll have to fill this property properly. If you need any help please let me know!

I'm having trouble understanding the point of the `providerNameSlug`. AFAICS, it's identical to the embed's `name` field [in all cases](https://github.com/WordPress/gutenberg/blob/e4931d8e536fb9ff69f0b9454e29139f9fe4ef5b/packages/block-library/src/embed/variations.js). Doesn't that make it redundant? (If it is, I'd suggest removing it, to avoid confusion)

Furthermore, does it work like it's supposed to? For example, if I embed any recent posts from ma.tt, the `providerNameSlug` I get from them is always `matt-mullenweg` rather than `wordpress`. :thinking:

**Comment by ntsekouras** [comment]
> Can you have a look what might've changed there? 🙏

`Preview` is retrieved from `getEmbedPreview core' selector and I haven't changed anything there. Since this is async, it certainly is `undefined` in some React lifecycles.

In time of development I had observed a weird behaviour of `WordPress` embed blocks and I had discussed it with @mcsf.  For example the `providerNameSlug` is indeed related to the site and not `wordpress` as I might have expected. Also related: https://github.com/WordPress/gutenberg/issues/21029.

We came to a conclusion not to alter any related functionality as that PR's main goal was the refactoring to a single block and not take a deep dive into `WordPress` embed block.

> I'm having trouble understanding the point of the providerNameSlug. AFAICS, it's identical to the embed's name field

For now it's the same but I don't know if it would be best to remove it. It feels like a `convention` to extract information from `name` for the `provider` property. Maybe in the future we'll have blocks from the same provider but different services, ex `youtube` could be the provider for more than one services? I'm not really sure about this though..

**Comment by ockham** [comment]
Thanks for getting back to me!

> > Can you have a look what might've changed there? pray
> 
> `Preview` is retrieved from `getEmbedPreview core' selector and I haven't changed anything there. Since this is async, it certainly is `undefined` in some React lifecycles.

The point is that it _always_ seems to be `undefined` for me now if I paste the URL of a WordPress site, whereas prior to this PR, it would indeed contain a preview of the given site. Can you repro this?

> In time of development I had observed a weird behaviour of `WordPress` embed blocks and I had discussed it with @mcsf. For example the `providerNameSlug` is indeed related to the site and not `wordpress` as I might have expected. Also related: #21029.
>
> We came to a conclusion not to alter any related functionality as that PR's main goal was the refactoring to a single block and not take a deep dive into `WordPress` embed block.

I'm a bit confused since I thought you [said](https://github.com/Automattic/wp-calypso/issues/39935#issuecomment-671825119) that the matching algorithm was changed to respect `providerNameSlug` per https://github.com/WordPress/gutenberg/pull/24090 :thinking: 

Overall, I agree on the approach of not making functional changes in a refactor PR -- I'm really just trying to find out what might've inadvertently broken the WordPress embed transformation.

> > I'm having trouble understanding the point of the providerNameSlug. AFAICS, it's identical to the embed's name field
> 
> For now it's the same but I don't know if it would be best to remove it. It feels like a `convention` to extract information from `name` for the `provider` property.

That makes sense for data that we fetch from 3rd party resources (like those embeds), but not so much for a hardcoded JSON, does it? :thinking: 

> Maybe in the future we'll have blocks from the same provider but different services, ex `youtube` could be the provider for more than one services? I'm not really sure about this though..

This seems like a textbook case of the YAGNI pattern -- we might need something in the future, but for now, the `providerNameSlug` is always identical to the `name`. I'd vote to remove -- if we need to add or modify functionality later, and that requires us to add a different field, we can do that then.

**Comment by ntsekouras** [comment]
> The point is that it always seems to be undefined for me now if I paste the URL of a WordPress site, whereas prior to this PR, it would indeed contain a preview of the given site. Can you repro this?

I can't because I had tried this when working on this PR and wasn't able to get `Preview` in the editor from master as well.
It seemed to be working on the front-end, both master and mine branch. Then the discussion happened not to dig into this, as we haven't changed the related functionality. I can give it a try again though.. Can you share a link that previously could fetch a `preview` and now can't?

> I'm a bit confused since I thought you said that the matching algorithm was changed to respect providerNameSlug per #24090 🤔

There was no change regarding the handling of `WordPress` embed block. The matching algorithm had to change as previously the other `embeds` were different blocks with different names. It doesn't affect the `preview` though from `getEmbedPreview` selector.

> That makes sense for data that we fetch from 3rd party resources (like those embeds), but not so much for a hardcoded JSON, does it? 🤔
> This seems like a textbook case of the YAGNI pattern

Yes it does and it overrides the property `conditionally` in `getAttributesFromPreview`. I'm not opposed to removing the attribute and it seems to be only used in `embed` block. The `conditionally` part should be checked though because there might be a need there for it after all, that is not clear to me at least. --cc @youknowriad @mcsf

**Comment by youknowriad** [comment]
Question here: What  happens to existing blocks? Do we keep them as hidden or something? Do we want to write a dev note about this change?

**Comment by ntsekouras** [comment]
> Question here: What happens to existing blocks? Do we keep them as hidden or something?

Backwards compatibility is handled in parser level in a similar fashion with `social block`: https://github.com/WordPress/gutenberg/blob/625136610f9fde02f2ff8b360cd811eb76da00a5/packages/blocks/src/api/parser.js#L441

>Do we want to write a dev note about this change?

A dev note about informing the change of `embed` block to a single block?

**Comment by youknowriad** [comment]
@ntsekouras yes, what happens if a user is using the old blocks in a CPT template for instance where the parser is not used?

**Comment by gziolo** [comment]
It would be great to expose `embed` block type on the server, similar to all other core blocks:
https://github.com/WordPress/gutenberg/blob/f724402a27106522a2fb94a4235fce4d5fc59632/lib/blocks.php#L16

We should also expose it in WordPress core, it needs to be added at least here:
https://github.com/WordPress/wordpress-develop/blob/701dee26f19b4bfdcbc1cc636f5fdecd8a753432/tools/webpack/packages.js#L118

**Comment by mcsf** [comment]
Good catch.
